source/azure_iot_provisioning_client.c

```diff
@@ -0,0 +1,963 @@
+                                                                    uint32_t ulIDScopeLength,
+                                                                    const uint8_t * pucRegistrationID,
+                                                                    uint32_t ulRegistrationIDLength,
+                                                                    AzureIoTProvisioningClientOptions_t * pxProvisioningClientOptions,
+                                                                    uint8_t * pucBuffer,
+                                                                    uint32_t ulBufferLength,
+                                                                    AzureIoTGetCurrentTimeFunc_t xGetTimeFunction,
+                                                                    const AzureIoTTransportInterface_t * pxTransportInterface )
+{
+    AzureIoTProvisioningClientResult_t xReturn;
+    az_span xEndpoint = az_span_create( ( uint8_t * ) pucEndpoint, ( int32_t ) ulEndpointLength );
+    az_span xIDScope = az_span_create( ( uint8_t * ) pucIDScope, ( int32_t ) ulIDScopeLength );
+    az_span xRegistrationID = az_span_create( ( uint8_t * ) pucRegistrationID, ( int32_t ) ulRegistrationIDLength );
+    az_result xCoreResult;
+    AzureIoTMQTTResult_t xResult;
+    az_iot_provisioning_client_options xOptions = az_iot_provisioning_client_options_default();
+
+    if( ( xAzureIoTProvisioningClientHandle == NULL ) ||
+        ( pucEndpoint == NULL ) || ( ulEndpointLength == 0 ) ||
+        ( pucIDScope == NULL ) || ( ulIDScopeLength == 0 ) ||
+        ( pucRegistrationID == NULL ) || ( ulRegistrationIDLength == 0 ) ||
+        ( pucBuffer == NULL ) || ( ulBufferLength == 0 ) ||
+        ( xGetTimeFunction == NULL ) || ( pxTransportInterface == NULL ) )
+    {
+        AZLogError( ( "Provisioning initialization failed: Invalid argument" ) );
+        xReturn = eAzureIoTProvisioningInvalidArgument;
+    }
+    else if( ( ulBufferLength < ( azureiotTOPIC_MAX + azureiotPROVISIONING_REQUEST_PAYLOAD_MAX ) ) ||
+             ( ulBufferLength < ( azureiotUSERNAME_MAX + azureiotPASSWORD_MAX ) ) )
+    {
+        AZLogError( ( "Provisioning initialize failed: Not enough memory passed \r\n" ) );
+        xReturn = eAzureIoTProvisioningOutOfMemory;
+    }
+    else
+    {
+        memset( xAzureIoTProvisioningClientHandle, 0, sizeof( AzureIoTProvisioningClient_t ) );
+        /* Setup scratch buffer to be used by middleware */
+        xAzureIoTProvisioningClientHandle->_internal.ulProvisioningScratchBufferLength =
+            azureMAX( ( azureiotUSERNAME_MAX + azureiotPASSWORD_MAX ),
+                      ( azureiotTOPIC_MAX + azureiotPROVISIONING_REQUEST_PAYLOAD_MAX ) );
+        xAzureIoTProvisioningClientHandle->_internal.pucProvisioningScratchBuffer = pucBuffer;
+        pucBuffer += xAzureIoTProvisioningClientHandle->_internal.ulProvisioningScratchBufferLength;
```

Comment:
Nit - consider a separate variable name for this, instead of re-using passed in parameter (compiler optimizers should reduce out in any case).  I think that makes a bit easier to read, like `pucNetworkBuffer` or something?